### PR TITLE
cannon: Rework RMW ops for 64-bit compatibility

### LINKED ISSUE #12242: ll/sc and lld/scd 64-bit emulation
Fix [`handleRMWOps`](https://github.com/ethereum-optimism/optimism/blob/dd2b21ce786f4c1b722bda270348597182153c8e/cannon/mipsevm/multithreaded/mips.go#L340) as it doesn't fully support 64-bits and may be out of spec (including possibly cross-incompatiblity between 32-bit and 64-bit reservations).

### PR DESCRIPTION
**Pull request by mbaxter**
**Description**

Rework RMW ops for 64-bit compatibility.

Changes include:
- `State.LLAddress` now holds the requested address rather than the effective address (which has the lower bits zeroed out).  It is necessary to retain some of the less significant bits in order to ensure that LL and SC operations are targeting the same 32-bit value when executing on 64-bit architectures.
- `bool State.LLReservationActive` has been changed to `uint8 State. LLReservationStatus`.  This allows us to store the type of reservation that was made (LL operations load a 32-bit value, LLD operations load a 64-bit value) so that the SC (32-bit) and SCD (64-bit) operations only succeed if a matching reservation has been made. 

**Tests**

- Updated existing ll, sc tests to work for 32- and 64-bit architectures
- Added 64-bit specific tests

**Metadata**
Fixes https://github.com/ethereum-optimism/optimism/issues/12242

### CODE REVIEW

**Comment by KADEFA2210** [comment]
Thankkk  adminnn .. goodluckk

Vào 05:20, T.6, 11 Th10, 2024 mbaxter ***@***.***> đã viết:

File: cannon/mipsevm/exec/mips_instructions.go [hunk 1]
```diff
@@ -569,3 +569,47 @@ func HandleRd(cpu *mipsevm.CpuScalars, registers *[32]Word, storeReg Word, val W
 	cpu.NextPC = cpu.NextPC + 4
 	return nil
 }
+
+func LoadSubWord(memory *memory.Memory, addr Word, byteLength Word, signExtend bool, memoryTracker MemTracker) Word {
```
Comment (mbaxter) [inline cannon/mipsevm/exec/mips_instructions.go:573]: These utils are probably only useful if we want to reuse them for other load/store operations.  On the plus side, I think it makes the code easier to reason about (esoteric bit manipulation is centralized and well-tested).  But it does make the code more complex in other ways.  Open to discussing.

File: cannon/mipsevm/memory/memory.go [hunk 2]
```diff
@@ -193,8 +194,8 @@ func (m *Memory) pageLookup(pageIndex Word) (*CachedPage, bool) {
 }
 
 func (m *Memory) SetUint32(addr Word, v uint32) {
-	// addr must be aligned to WordSizeBytes bytes
-	if addr&arch.ExtMask != 0 {
+	// addr must be aligned to 4-byte (32-bit) boundaries
+	if addr&3 != 0 {
```
Comment (mbaxter) [inline cannon/mipsevm/memory/memory.go]: Calling out this unrelated change here - I think we want to be able to set uint32's at any even multiple of 4 bytes.

File: cannon/mipsevm/testutil/mips.go [hunk 3]
```diff
@@ -172,6 +172,12 @@ func LogStepFailureAtCleanup(t *testing.T, mipsEvm *MIPSEVM) {
 
 // ValidateEVM runs a single evm step and validates against an FPVM poststate
 func ValidateEVM(t *testing.T, stepWitness *mipsevm.StepWitness, step uint64, goVm mipsevm.FPVM, hashFn mipsevm.HashFn, contracts *ContractMetadata, tracer *tracing.Hooks) {
+	if !arch.IsMips32 {
```
Comment (mbaxter) [inline cannon/mipsevm/testutil/mips.go:175]: Disabled EVM execution for differential tests on 64-bit architecture for the time being.

File: cannon/mipsevm/multithreaded/mips.go [hunk 4]
```diff
@@ -343,36 +343,40 @@ func (m *InstrumentedState) handleRMWOps(insn, opcode uint32) error {
...
+
+		m.state.LLReservationStatus = targetStatus
+		m.state.LLAddress = addr
 		m.state.LLOwnerThread = threadId
-	} else if opcode == exec.OpStoreConditional || opcode == exec.OpStoreConditional64 {
-		// TODO(#12205): Determine bits affected by coherence stores on 64-bits
-		// Check if our memory reservation is still intact
-		if m.state.LLReservationActive && m.state.LLOwnerThread == threadId && m.state.LLAddress == effAddr {
+	case exec.OpStoreConditional, exec.OpStoreConditional64:
+		if m.state.LLReservationStatus == targetStatus && m.state.LLOwnerThread == threadId && m.state.LLAddress == addr {
```
Comment (Inphi) [inline cannon/mipsevm/multithreaded/mips.go:367]: The store is now conditional on the virtual address. I think we should use the effective address instead because the `LLAddress` is cleared based on that.
Otherwise, if `LLAddress=0x101` and a regular store to 0x100 occurred, then we `sc[d]` will erroneously succeed. This scenario may be contrived though, since the vaddr should already be naturally aligned (to 4 and 8 bytes for sc and scd) and I don't expect compilers violate this alignment constraint. Nevertheless, it's good to be extra defensive here.

File: cannon/mipsevm/memory/memory.go (hunk 2 above)
Comment (Inphi) [reply cannon/mipsevm/memory/memory.go]: The only usecase I can think of to do this is for writing 4-byte instructions at 4-byte boundaries. I'd prefer that those callsites should be modified to use 8-byte aligned addresses.
Is any other usecase for this for the 64-bit VM? Relaxing the alignment constraint here breaks my mental model of how stores are emulated for a 64-bit CPU.

Comment (mbaxter) [reply cannon/mipsevm/memory/memory.go]: I think this method is exclusively used in tests.  So we should probably just cut this from the production code altogether, and make a test util or else migrate the existing calls to use 8-byte boundaries as you suggest.  I can revert my change for now and add a ticket to clean this up.

File: cannon/mipsevm/multithreaded/mips.go (hunk 4 above)
Comment (mbaxter) [reply cannon/mipsevm/multithreaded/mips.go:367]: We now mask the `LLAddress` with `arch.AddressMask` [here](https://github.com/ethereum-optimism/optimism/blob/de438d3ad389b61603dac01e230e232f77b29e42/cannon/mipsevm/multithreaded/mips.go#L328) when checking if a reservation should be invalidated.  So I don't think we'll fail to invalidate unless I'm missing something?  We could defensively mask both addresses when checking, but the `memAddr` value there should already be masked.

The reserved address is a little complicated in 64-bit MIPS.  For the `ll`/`sc` instructions, we want to make sure the 4-byte address offset matches between load and store.  But if we just store the effective address to `LLAddress`, that offset bit will be lost.

Or are you suggesting we mask the LLAddress based on the operation type?  So, for LL/SC, we store a 4-byte aligned value and for LLD/SCD, we store an 8-byte aligned value?  I think that could work, but it seems to complicate the code without giving us much benefit since we need to mask it when we do the memory check anyway. 

There's also this in the MIPS `sc` documentation, which implies we should be doing stricter checks:
> A RMW sequence executed without intervening exceptions must use the same address in the LL and SC.The address is the same if the virtual address, physical address, and cache-coherence algorithm are identical.

File: cannon/mipsevm/memory/memory.go (hunk 2 above)
Comment (Inphi) [reply cannon/mipsevm/memory/memory.go]: +1 on removing this from production code. We have delegate it to a testutil helper function that only sets up the instruction word.

File: cannon/mipsevm/multithreaded/mips.go (hunk 4 above)
Comment (Inphi) [reply cannon/mipsevm/multithreaded/mips.go:367]: > Or are you suggesting we mask the LLAddress based on the operation type? So, for LL/SC, we store a 4-byte aligned value and for LLD/SCD, we store an 8-byte aligned value? I think that could work, but it seems to complicate the code without giving us much benefit since we need to mask it when we do the memory check anyway.

I think so because the current mask breaks single-word ll/sc semantics. Consider when a RMW sequence is initiated for a lower word (ex: 0x104) and a 32-bit instruction issues a write to the bits between [0x100, 0x104). A subsequent `sc` would succeed on a 32-bit CPU, so I'd expect the same for the 64-bit CPU per backwards compatibility. 
```
ll rt, 0x100(4)
sw rt, 0x100(0)
sc rt, 0x100(4)
```
However, `handleMemoryUpdate` clears the memory reserved at 0x104 since it's located on the same effective address.

Comment (Inphi) [reply cannon/mipsevm/multithreaded/mips.go:367]: This might be OK. The spec has this to say about such stores:
> If either of the following events occurs between the execution of LL and SC, the SC fails:
> A coherent store is completed by another processor or coherent I/O module into the block of synchronizable
physical memory containing the word. The size and alignment of the block is implementation-dependent, but it is at least one word and at most the minimum page size.

So technically, cannon64 would be within spec. It's just that it results spurious sc failures if a store occurs within the same doubelword block. In which case, I'd optimize for implementation simplicity like you suggested rather than try to eliminate these spurious sc failures.

File: cannon/mipsevm/tests/evm_multithreaded_test.go [hunk 5]
```diff
@@ -94,38 +104,49 @@ func TestEVM_MT_SC(t *testing.T) {
...
+		{name: "Zero valued ll args", base: 0x00_00_00_00, offset: 0x0, value: 0xABCD, rtReg: 5, threadId: 0},
 	}
 	for i, c := range cases {
 		for _, v := range llVariations {
 			tName := fmt.Sprintf("%v (%v)", c.name, v.name)
 			t.Run(tName, func(t *testing.T) {
+				// Perform some calculations for 64-bit compatibility
+				// Calculate address
+				addrOffset, signExtended := testutil.SignExtendImmediate(Word(c.offset))
+				require.Equal(t, signExtended, c.shouldSignExtendOffset)
```
Comment (Inphi) [inline cannon/mipsevm/tests/evm_multithreaded_test.go]: Is this just a sanity check on the test case inputs? Which is OK to do, but I don't see `shouldSignExtendOffset` used anywhere else.

File: cannon/mipsevm/tests/evm_multithreaded_test.go [hunk 6]
```diff
@@ -94,38 +104,49 @@ func TestEVM_MT_SC(t *testing.T) {
...
 		shouldSucceed       bool
 	}{
-		{name: "should succeed", llReservationActive: true, matchThreadId: true, matchEffAddr: true, shouldSucceed: true},
-		{name: "mismatch addr", llReservationActive: true, matchThreadId: false, matchEffAddr: true, shouldSucceed: false},
-		{name: "mismatched thread", llReservationActive: true, matchThreadId: true, matchEffAddr: false, shouldSucceed: false},
-		{name: "mismatched addr & thread", llReservationActive: true, matchThreadId: false, matchEffAddr: false, shouldSucceed: false},
-		{name: "no active reservation", llReservationActive: false, matchThreadId: true, matchEffAddr: true, shouldSucceed: false},
+		{name: "should succeed", llReservationStatus: multithreaded.LLStatusActive32bit, matchThreadId: true, matchAddr: true, shouldSucceed: true},
+		{name: "mismatch addr", llReservationStatus: multithreaded.LLStatusActive32bit, matchThreadId: false, matchAddr: true, shouldSucceed: false},
+		{name: "mismatched thread", llReservationStatus: multithreaded.LLStatusActive32bit, matchThreadId: true, matchAddr: false, shouldSucceed: false},
```
Comment (Inphi) [inline cannon/mipsevm/tests/evm_multithreaded_test.go]:
```suggestion
		{name: "mismatch thread", llReservationStatus: multithreaded.LLStatusActive32bit, matchThreadId: false, matchAddr: true, shouldSucceed: false},
		{name: "mismatched addr", llReservationStatus: multithreaded.LLStatusActive32bit, matchThreadId: true, matchAddr: false, shouldSucceed: false},
```

File: cannon/mipsevm/memory/memory.go (hunk 2 above)
Comment (mbaxter) [reply cannon/mipsevm/memory/memory.go]: Added a ticket [here](https://github.com/ethereum-optimism/optimism/issues/12488).  Should I add another ticket to cut `memory.GetUint32`?

Comment (Inphi) [reply cannon/mipsevm/memory/memory.go]: Thanks for creating the ticket. That issue can tackle both.

**Review by Inphi — APPROVED** [review]
LGTM! I appreciate the test for backwards-compatibility for 32-bit instructions on cannon64.
Just a couple nits.

File: cannon/mipsevm/tests/evm_multithreaded_test.go (hunk 5 above)
Comment (mbaxter) [reply cannon/mipsevm/tests/evm_multithreaded_test.go]: Yeah, that was a sanity check, but I pushed a few more commits and reworked this, so its gone now.
